### PR TITLE
MAINT: Read requirements from setup.py

### PR DESCRIPTION
**Pull request by ssanderson**
@llllllllll for you :heart:.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -14,32 +14,101 @@
...
+            continue
+
+        # pip install -r understands line with ;python_version<'3.0', but
+        # whatever happens inside extras_requires doesn't.  Parse the line
+        # manually and conditionally add it if needed.
+        if ';' in line:
+            requirement, version_spec = line.split(';')
+            try:
+                groups = re.match(
+                    "(python_version)([<>=]+)(')([0-9\.]+)(')(.*)",
```
Comment (llllllllll) [inline setup.py]: `\2` should be `[<>=]{1, 2}` or we get things like js super equals `===`

File: setup.py [hunk 2]
```diff
@@ -14,32 +14,101 @@
...
+                    version_spec,
+                ).groups()
+                comp = {'<': lt, '=': eq, '>': gt}[groups[1]]
+                version_spec = StrictVersion(groups[3])
+            except Exception as e:
+                # My kingdom for a 'raise from'!
```
Comment (llllllllll) [inline setup.py]: you can emulate this in python 2 by raising a tuple of `(value, None, tb)`

File: setup.py [hunk 3]
```diff
@@ -14,32 +14,101 @@
...
+def _filter_requirements(lines_iter):
+    for line in lines_iter:
+        line = line.strip()
+        if not line or line.startswith('#'):
+            continue
+
+        # pip install -r understands line with ;python_version<'3.0', but
+        # whatever happens inside extras_requires doesn't.  Parse the line
+        # manually and conditionally add it if needed.
+        if ';' in line:
```
Comment (llllllllll) [inline setup.py]: does it need to start with this?

File: setup.py [hunk 4]
```diff
@@ -14,32 +14,101 @@
...
+            requirement, version_spec = line.split(';')
+            try:
+                groups = re.match(
+                    "(python_version)([<>=]+)(')([0-9\.]+)(')(.*)",
+                    version_spec,
+                ).groups()
+                comp = {'<': lt, '=': eq, '>': gt}[groups[1]]
```
Comment (llllllllll) [inline setup.py]: what about `>=, `==', `<=`, ...

File: setup.py [hunk 5]
```diff
@@ -14,32 +14,101 @@
...
+                version_spec = StrictVersion(groups[3])
+            except Exception as e:
+                # My kingdom for a 'raise from'!
+                raise ValueError(
+                    "Couldn't parse requirement line; '%s'\n"
+                    "Error was:\n"
+                    "%r" % (line, e)
+                )
+
+            sys_version = '.'.join(list(map(str, sys.version_info[:3])))
```
Comment (llllllllll) [inline setup.py]: join accepts an iterable

File: setup.py [hunk 6]
```diff
@@ -14,32 +14,101 @@
...
+    abspath,
+    dirname,
+    join,
+)
+from distutils.version import StrictVersion
 from setuptools import setup, Extension
-from Cython.Build import cythonize
-import numpy as np
+try:
+    from Cython.Build import cythonize
```
Comment (llllllllll) [inline setup.py]: How are we putting this on pypi now?

**Comment by llllllllll** [comment]
This is a good change, just had a few comments / questions.

File: setup.py (hunk 3 above)
Comment (ssanderson) [inline setup.py]: The line that this is for is `backports.ssl-match-hostname==3.4.0.2;python_version<'3.0'`.

Comment (llllllllll) [inline setup.py]: kk

**Pull request closed without merging**